sdks/python/apache_beam/transforms/environments.py

```diff
@@ -0,0 +1,331 @@
+                         beam_runner_api_pb2.ProcessPayload, ProcessEnvironment)
+
+
+class ExternalEnvironment(Environment):
+
+  def __init__(self, url, params=None):
+    self.url = url
+    self.params = params
+
+  def __eq__(self, other):
+    return self.__class__ == other.__class__ and self.url == other.url \
+      and self.params == other.params
+
+  def __ne__(self, other):
+    # TODO(BEAM-5949): Needed for Python 2 compatibility.
+    return not self == other
+
+  def __hash__(self):
+    return hash((self.__class__, self.url, self.params))
+
+  def __repr__(self):
+    return 'ExternalEnvironment(url=%s,params=%s)' % (self.url, self.params)
+
+  def to_runner_api_parameter(self, context):
+    return (common_urns.environments.EXTERNAL.urn,
+            beam_runner_api_pb2.ExternalPayload(
+                endpoint=endpoints_pb2.ApiServiceDescriptor(url=self.url),
+                params=self.params
+            ))
+
+  @staticmethod
+  def from_runner_api_parameter(payload, context):
+    return ExternalEnvironment(payload.endpoint.url,
+                               params=payload.params or None)
+
+  @classmethod
+  def from_options(cls, options):
+    def looks_like_json(environment_config):
+      import re
+      return re.match(r'\s*\{.*\}\s*$', environment_config)
+
+    if looks_like_json(options.environment_config):
+      config = json.loads(options.environment_config)
+      url = config.get('url')
+      if not url:
+        raise ValueError('External environment endpoint must be set.')
+      params = config.get('params')
+    else:
+      url = options.environment_config
+      params = None
+
+    return cls(url, params=params)
+
+
+Environment.register_urn(common_urns.environments.EXTERNAL.urn,
+                         beam_runner_api_pb2.ExternalPayload,
+                         ExternalEnvironment)
+
+
+class PythonSDKEnvironment(Environment):
+
+  def __init__(self, payload=None):
+    self.payload = payload
+
+  def __eq__(self, other):
+    return self.__class__ == other.__class__ and self.payload == other.payload
+
+  def __ne__(self, other):
+    # TODO(BEAM-5949): Needed for Python 2 compatibility.
+    return not self == other
+
+  def __hash__(self):
+    return hash((self.__class__, self.payload))
+
+  def __repr__(self):
+    return '%s(payload=%s)' % (self.__class__.__name__, self.payload)
+
+  @classmethod
+  def from_options(cls, options):
+    return cls(payload=options.environment_config)
+
+
+@Environment.register_python_environment(python_urns.EMBEDDED_PYTHON, bytes)
+class EmbeddedPythonEnvironment(PythonSDKEnvironment):
+  pass
+
+
+@Environment.register_python_environment(python_urns.EMBEDDED_PYTHON_GRPC,
+                                         bytes)
+class EmbeddedPythonGrpcEnvironment(PythonSDKEnvironment):
+  pass
+
+
+@Environment.register_python_environment(python_urns.SUBPROCESS_SDK, bytes)
+class SubprocessSDKEnvironment(PythonSDKEnvironment):
+  pass
```

Comment:
I have an idea to further simplify this, based on this observation:  "EMBEDDED_PYTHON_GRPC is the only of these python environments that actually has a payload".

Here's roughly what I would do.  

```python

@Environment.register_urn(python_urns.EMBEDDED_PYTHON, None)
class EmbeddedPythonEnvironment(Environment):
  def to_runner_api_parameter(self, context):
    return (common_urns.environments.EMBEDDED_PYTHON.urn, None)

  @staticmethod
  def from_runner_api_parameter(unused_payload, context):
    # parse payload into args
    return EmbeddedPythonEnvironment()


@Environment.register_python_environment(python_urns.EMBEDDED_PYTHON_GRPC,
                                         bytes)
class EmbeddedPythonGrpcEnvironment(Environment):
  def __init__(num_workers=None, state_cache_size=None):
      self.num_workers = num_workers
      self.state_cache_size = state_cache_size

  def to_runner_api_parameter(self, context):
    if self.num_workers and self.state_cache_size:
      payload = b'%d%d' % (self.num_workers, self.state_cache_size)
    elif:
      raise ???
    else:
      payload = b''

    return (common_urns.environments.DOCKER.urn, payload)

  @staticmethod
  def from_runner_api_parameter(payload, context):
    # parse payload into args
    return EmbeddedPythonGrpcEnvironment(num_workers=x, state_cache_size=y)


@Environment.register_python_environment(python_urns.SUBPROCESS_SDK, bytes)
class SubprocessSDKEnvironment(Environment):
  def __init__(self, command_string):
    self.command_string = command_string

  def to_runner_api_parameter(self, context):
    return (common_urns.environments.SUBPROCESS_SDK.urn, self.command_string.encode('utf-8'))

  @staticmethod
  def from_runner_api_parameter(payload, context):
    # parse payload into args
    return SubprocessSDKEnvironment(payload.decode('utf-8'))
```

Reply:
Updated in https://github.com/violalyu/beam/pull/1/commits/45e955a6de7c944a1b540987d59217ab627ddab9